### PR TITLE
[metricbeat][Kubernetes] Update token on the first HTTP 401 when making requests to kubelet API

### PR DESCRIPTION
**Pull request by constanca-m**
## What & How

Please read https://github.com/elastic/beats/issues/40627.

I also added comments on the code explaining what is happening. A quick summary:

- The token used to authenticate requests to the Kubelet can expire. If this happens, we will get a HTTP 401 error. In this case, we can update the Authorization header again by reading the file and updating the header with the new value. This is only done once. We do this because the HTTP 401 error could be caused by something else (I do not see anything else that could cause this, but just be sure), and in that case we do not want to get stuck in a loop of trying to fetch content that is not available.

## Issues

Closes https://github.com/elastic/beats/issues/40627.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test & Results

1. You can use [dev-tools](https://github.com/elastic/beats/tree/main/dev-tools/kubernetes) to test this. You can change the metricbeat manifest to include this:

```yaml
    # In the config map to enable the metricsets:
      - module: kubernetes
        metricsets:
          - node
          - system
          - pod
          - container
          - volume
        period: 10s
        host: ${NODE_NAME}
        hosts: ["https://${NODE_NAME}:10250"]
        bearer_token_file: /service-account/token
        ssl.verification_mode: "none"


      # In your metricbeat pod:
        [...]
            volumeMounts:
            - name: token-vol
              mountPath: /service-account
              readOnly: true
        [...]
        volumes:
          - name: token-vol
            projected:
              sources:
                - serviceAccountToken:
                    path: token
                    expirationSeconds: 600
```

This means that the token used to authenticate the requests expires after 10min. This is the minimum time we can wait to refresh tokens.
From [Bound service account token volume mechanism](https://kubernetes.io/docs/reference/access-authn-authz/service-accounts-admin/):
> Note:
**There is no specific mechanism to invalidate a token** (...)

2. After waiting 10 minutes:
    1. If your using the **current metricbeat**: you should see in the logs:
    ```
   {"log.level":"error","@timestamp":"2024-08-28T07:27:28.507Z","log.logger":"kubernetes.container","log.origin":{"function":"github.com/elastic/beats/v7/metricbeat/module/kubernetes/container.(*MetricSet).Fetch","file.name":"container/container.go","file.line":95},"message":"HTTP error 401 in : 401 Unauthorized","service.name":"metricbeat","ecs.version":"1.6.0"}
   ```
    2. If your using the **metricbeat from this PR**: There is no error in the logs. The token gets updated and metrics continue being successfully fetched.

### CODE REVIEW

File: metricbeat/helper/http.go [hunk 1]
```diff
@@ -124,10 +127,16 @@ func (h *HTTP) FetchResponse() (*http.Response, error) {
...
+	fmt.Println(">>>> Making a request with HEADERS:")
+	for k, v := range req.Header {
+		fmt.Println(">>>> ", k, " : ", v)
+	}
 	resp, err := h.client.Do(req)
 	if err != nil {
+		fmt.Println(">>>>>> Request failed.", err)
 		return nil, fmt.Errorf("error making http request: %v", err)
 	}
+	fmt.Println(">>>>>> Request succeeded.")
```
Comment (gizas) [inline metricbeat/helper/http.go:139]: these println should be removed right?

File: metricbeat/helper/http_test.go [hunk 2]
```diff
@@ -295,6 +296,42 @@ func TestUserAgentCheck(t *testing.T) {
...
+	assert.NoError(t, err)
+
+	helper := &HTTP{bearerFile: bearerFilePath, headers: make(http.Header)}
+	updated, err := helper.RefreshAuthorizationHeader()
+	assert.NoError(t, err)
+	assert.True(t, updated)
+	expected := fmt.Sprintf("Bearer %s", firstToken)
+	assert.Equal(t, expected, getAuth(helper))
+
+	secondToken := "token-1"
```
Comment (gizas) [inline metricbeat/helper/http_test.go:324]: we should change it token-2 and not have the same value

File: metricbeat/module/kubernetes/kubernetes.go [hunk 3]
```diff
@@ -142,13 +144,32 @@ func (m *module) GetKubeletStats(http *helper.HTTP) ([]byte, error) {
...
 
-	// NOTE: These entries will be never removed, this can be a leak if
+	// NOTE: These entries will never be removed, this can be a leak if
 	// metricbeat is used to monitor clusters dynamically created.
 	// (https://github.com/elastic/beats/pull/25640#discussion_r633395213)
 	statsCache := m.kubeletStatsCache.getCacheMapEntry(m.cacheHash)
 
+	// Check if the last time we tried to make a request to the Kubelet API ended in a 401 Unauthorized error.
+	// If this is the case, we should not keep making requests.
+	errorMsg := fmt.Sprintf(fmt.Sprintf("HTTP error %d", httpnet.StatusUnauthorized))
```
Comment (gizas) [inline metricbeat/module/kubernetes/kubernetes.go:154]:
```suggestion
	errorUnathorisedMsg := fmt.Sprintf(fmt.Sprintf("HTTP error %d", httpnet.StatusUnauthorized))
```

File: metricbeat/helper/http.go (hunk 1 above)
Comment (constanca-m) [reply metricbeat/helper/http.go]: Ahah yes! I knew I had left these somewhere...

**Comment by gizas** [comment]
Really nice work Constanca!

**Review by gizas — APPROVED** [review]

**Comment by mauri870** [comment]
> lint error is a mistake. The `NewRequest` function already uses `NewRequestWithContext`, which is the lint suggestion:
> 
> ```go
> // NewRequest wraps [NewRequestWithContext] using [context.Background].
> func NewRequest(method, url string, body io.Reader) (*Request, error) {
> 	return NewRequestWithContext(context.Background(), method, url, body)
> }
> ```
> 
> Source: https://github.com/golang/go/blob/9e8ea567c838574a0f14538c0bbbd83c3215aa55/src/net/http/request.go#L862-L864

I believe this behavior is intentional. While NewRequest internally calls NewRequestWithContext, it abstracts away the use of context so that you don't need to worry about it. The general consensus for context-aware functions is that it's better to accept a context as the first parameter, even if you don't plan to use it immediately. This approach makes it clear and straightforward to add a context later when it's available.

In this case, the higher-level method doesn't accept a context, but I still think it's better to be explicit and use NewRequestWithContext regardless. My recommendation is to use http.NewRequestWithContext(context.TODO(), ...).

**Review by mauri870 — APPROVED** [review]

**Pull request merged**